src/main/scala/lms/thirdparty/mpi.scala

```diff
@@ -0,0 +1,87 @@
+package lms.thirdparty
+
+import lms.core._
+import lms.util._
+import lms.core.stub._
+import lms.core.Backend._
+import lms.core.virtualize
+import lms.core.utils.time
+import lms.macros.SourceContext
+
+import lms.collection._
+
+trait MPIOps extends CMacro with LibStruct { b: Base =>
+  /* LMS support for MPI library */
+
+  // this is how we deal with constant macros
+  class MPIWorld
+  def mpi_comm_world: Rep[MPIWorld] = cmacro[MPIWorld]("MPI_COMM_WORLD")
+  def mpi_max_processor_name: Rep[Int] = cmacro[Int]("MPI_MAX_PROCESSOR_NAME")
+
+  // this is how we deal with library functions (may need pointers)
+  def mpi_init(): Rep[Unit] = Wrap[Unit](Adapter.g.reflectWrite("mpi-init")(Adapter.CTRL))
+  def mpi_finalize(): Rep[Unit] = Wrap[Unit](Adapter.g.reflectWrite("mpi-finalize")(Adapter.CTRL))
+
+  def mpi_comm_size(world: Rep[MPIWorld], size: Var[Int]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("mpi-comm-size", Unwrap(world), UnwrapV(size))(UnwrapV(size)))
+  }
+
+  def mpi_comm_rank(world: Rep[MPIWorld], rank: Var[Int]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("mpi-comm-rank", Unwrap(world), UnwrapV(rank))(UnwrapV(rank)))
+  }
+
+  def mpi_get_processor_name(name: Rep[Array[Char]], len: Var[Int]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("mpi-get-processor-name", Unwrap(name), UnwrapV(len))
+    (Unwrap(name), UnwrapV(len)))
+  }
+
+  // this is how we bind to library structs
+  abstract class DataStructure1
+  def dataStructure1 = libStruct[DataStructure1]("DataStructure1")
+
+  // some example (dummy code). Please remove later :)
+  def test_pointer(x: Var[Int]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("test-pointer", UnwrapV(x))(Adapter.CTRL, UnwrapV(x)))
+  }
+  def test_pointer(x: Rep[Array[Int]]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("test-pointer-array", Unwrap(x))(Adapter.CTRL, Unwrap(x)))
+  }
+  def test_struct_pointer(x: Rep[DataStructure1]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("test-struct-pointer", Unwrap(x))(Adapter.CTRL, Unwrap(x)))
+  }
+  def test_struct_reference(x: Rep[DataStructure1]): Rep[Unit] = {
+    Wrap[Unit](Adapter.g.reflectWrite("test-struct-ref", Unwrap(x))(Adapter.CTRL, Unwrap(x)))
+  }
+}
+
+trait CCodeGenMPI extends ExtendedCCodeGen {
+  override def remap(m: Manifest[_]): String =
+    m.runtimeClass.getName match {
+      case "lms.thirdparty.MPIOps$DataStructure1" => "DataStructure1"
```

Comment:
this case of `remap` is still needed, but it is simply because one of the test defines a new `topFun` using the type of the `DataStruct1`. This rarely happens in library bindings :)